packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx

```diff
@@ -135,7 +135,53 @@ export const useGridVirtualScroller = () => {
   const columnsTotalWidth = dimensions.columnsTotalWidth;
   const hasColSpan = useGridSelector(apiRef, gridHasColSpanSelector);
 
-  useResizeObserver(mainRef, () => apiRef.current.resize());
+  const mainRefCallback = React.useCallback(
+    (node: HTMLDivElement | null) => {
+      mainRef.current = node;
+
+      if (!node || typeof ResizeObserver === 'undefined') {
+        return undefined;
+      }
+
+      const initialRect = node.getBoundingClientRect();
+      let lastSize = {
+        width: initialRect.width,
+        height: initialRect.height,
+      };
+
+      apiRef.current.publishEvent('resize', lastSize);
+
+      const observer = new ResizeObserver((entries) => {
+        const entry = entries[0];
+        if (!entry) {
+          return;
+        }
+
+        const newSize = {
+          width: entry.contentRect.width,
+          height: entry.contentRect.height,
+        };
+
+        if (newSize.width === lastSize.width && newSize.height === lastSize.height) {
+          return;
+        }
+
+        apiRef.current.publishEvent('resize', newSize);
+        lastSize = newSize;
+      });
+
+      observer.observe(node);
+
+      if (reactMajor >= 19) {
+        return () => {
+          mainRef.current = null;
+          observer.disconnect();
+        };
+      }
+      return undefined;
```

Comment:
Is this really correct? We're not disconnecting on react <= 18?

Reply:
Right, ref cleanup is only supported in React 19.
Also, I'm checking the React major because React 18 logs this error: https://github.com/facebook/react/blob/f1338f8080abd1386454a10bbf93d67bfe37ce85/packages/react-reconciler/src/ReactFiberCommitWork.new.js#L292-L295
I can add a cleanup effect for older versions.

Reply:
This issue wasn't addressed, I think this might be leaking memory.

Reply:
Yeah, I forgot about this one.
I'll open a PR with a cleanup effect.

Reply:
Btw I would skip the version checking and use a single strategy, to keep the code simple.

Reply:
Tested in Chrome/Edge, and no detached node is left behind after unmounting.

Run this:
const observer = new ResizeObserver(()=>true);
observer.observe(document.querySelector(".MuiDataGrid-root"));

Make the datagrid unmount. Verify that no detached nodes are left behind.

Reply:
> Verify that no detached nodes are left behind.

How do you verify this?
I see retained detached nodes in Chrome Devtools, even with the resize observer commented out in `useGridVirtualScroller`:


https://github.com/user-attachments/assets/6a50272e-8a6d-493d-b2de-f4910b13f715

Reply:
Try here: https://w8n3kg.csb.app/
Source: https://codesandbox.io/p/sandbox/w8n3kg

Whole Datagrid was a bad example perhaps. There may or may not be a memory leak somewhere else there. I've spent good deal of time trying to debug this in our own app, and I'm not sure at this point. Could be that it's very slow to garbage collect (which I don't understand why – (is it the DOM tree size?) and running a profile too quickly creates a reference to the detached node, which in turn prevents garbage collection). But not sure either if and when it would automatically get garbage collected. Bugs in devotools doesn't make it a joy to debug.

See: https://github.com/mui/mui-x/issues/15459#issuecomment-2497922244

Reply:
Are you confident that everything is GC'ed correctly across all browsers? I would tend to add cleanup just to be safe.

Reply:
Can never be sure about that, but from quick checks with Safari and Firefox, it seems to be the case. But let's add it to be extra safe.